[macOS] Scrollbar corners fail to render with UI-side compositing

pxlcoder opened the pull request:
#### 44477657ccd98fcf58206f2aa87b24ef84556d3a
<pre>
[macOS] Scrollbar corners fail to render with UI-side compositing
<a href="https://bugs.webkit.org/show_bug.cgi?id=248763">https://bugs.webkit.org/show_bug.cgi?id=248763</a>
rdar://102906463

Reviewed by Simon Fraser.

Painting scrollbar corners is broken due to the lack of a platform context in
the web process. To fix, delegate drawing to the GPU process by using
`drawSystemImage` and introducing new `SystemImage` subclasses.

Scrollbar corners are currently drawn using AppKit and CoreUI.
`AppKitControlSystemImage` is introduced to provide a generic method of
rendering native controls, encapsulating the desired color scheme and tint
color. `ScrollbarTrackCornerSystemImageMac` is the concrete subclass
responsible for drawing the native control. Additional subclasses will be added
in subsequent patches, to support other controls that are drawn using CoreUI
options (progress bars and datalist dropdown buttons).

* Source/WebCore/PAL/pal/spi/mac/NSAppearanceSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/SystemImage.h:
* Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.h: Added.
(isType):
* Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.mm: Added. Source/WebCore/platform/graphics/SystemImage.h.
(WebCore::AppKitControlSystemImage::AppKitControlSystemImage):
(WebCore::AppKitControlSystemImage::draw const):
* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.h: Added.
(isType):
* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.mm: Added.
(WebCore::ScrollbarTrackCornerSystemImageMac::ScrollbarTrackCornerSystemImageMac):
(WebCore::ScrollbarTrackCornerSystemImageMac::drawControl const):
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::paintScrollCorner):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;SystemImage&gt;::encode):
(IPC::ArgumentCoder&lt;SystemImage&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder&lt;WebCore::AppKitControlSystemImage&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::AppKitControlSystemImage&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/257421@main">https://commits.webkit.org/257421@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6d555408cc7a0bbb85d5acf6fcd13d4efae4456

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98864 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108279 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168536 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102819 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85442 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91397 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106265 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90104 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33566 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21448 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76423 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1986 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22977 "Found 30 new test failures: http/tests/appcache/whitelist-wildcard.html, http/tests/cache/disk-cache/shattered-deduplication.html, http/tests/security/mixedContent/secure-redirect-to-secure-redirect-to-basic-auth-insecure-image.https.html, http/wpt/cache-storage/cache-quota.any.html, http/wpt/fetch/response-opaque-clone.html, imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-sends-reports-on-violation.https.sub.html, imported/w3c/web-platform-tests/css/css-flexbox/flex-outer-flexbox-column-recalculate-height-on-resize-001.html, imported/w3c/web-platform-tests/css/cssom-view/HTMLBody-ScrollArea_quirksmode.html, imported/w3c/web-platform-tests/feature-policy/reporting/microphone-reporting.https.html, imported/w3c/web-platform-tests/html/cross-origin-embedder-policy/require-corp-load-from-cache-storage.https.html ... (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1895 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45433 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5107 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6851 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42437 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3294 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.h:
```diff
@@ -0,0 +1,87 @@
+private:
+    AppKitControlSystemImageType m_controlType;
+
+    bool m_useDarkAppearance { false };
```
Probably better for padding if the bool comes last.

smfr commented on Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
```diff
@@ -271,4 +274,56 @@ static void encodeNSError(Encoder& encoder, NSError *nsError)
 #endif
 
+template<typename Encoder>
+void ArgumentCoder<WebCore::AppKitControlSystemImage>::encode(Encoder& encoder, const WebCore::AppKitControlSystemImage& systemImage)
```
Can we use Alex's new generated encoding/decoding thing here?

pxlcoder replied:
The new mechanism doesn't support switching based on derived classes yet.

However, we can use the new approach to implement encoding/decoding of `AppKitControlSystemImage` subclasses. The one added in this patch doesn't have additional members, but the one to support `<progress>` will.

pxlcoder committed:
[macOS] Scrollbar corners fail to render with UI-side compositing
https://bugs.webkit.org/show_bug.cgi?id=248763
rdar://102906463

Reviewed by Simon Fraser.

Painting scrollbar corners is broken due to the lack of a platform context in
the web process. To fix, delegate drawing to the GPU process by using
`drawSystemImage` and introducing new `SystemImage` subclasses.

Scrollbar corners are currently drawn using AppKit and CoreUI.
`AppKitControlSystemImage` is introduced to provide a generic method of
rendering native controls, encapsulating the desired color scheme and tint
color. `ScrollbarTrackCornerSystemImageMac` is the concrete subclass
responsible for drawing the native control. Additional subclasses will be added
in subsequent patches, to support other controls that are drawn using CoreUI
options (progress bars and datalist dropdown buttons).

* Source/WebCore/PAL/pal/spi/mac/NSAppearanceSPI.h:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/SystemImage.h:
* Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.h: Added.
(isType):
* Source/WebCore/platform/graphics/mac/AppKitControlSystemImage.mm: Added. Source/WebCore/platform/graphics/SystemImage.h.
(WebCore::AppKitControlSystemImage::AppKitControlSystemImage):
(WebCore::AppKitControlSystemImage::draw const):
* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.h: Added.
(isType):
* Source/WebCore/platform/graphics/mac/ScrollbarTrackCornerSystemImageMac.mm: Added.
(WebCore::ScrollbarTrackCornerSystemImageMac::ScrollbarTrackCornerSystemImageMac):
(WebCore::ScrollbarTrackCornerSystemImageMac::drawControl const):
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::paintScrollCorner):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<SystemImage>::encode):
(IPC::ArgumentCoder<SystemImage>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/mac/WebCoreArgumentCodersMac.mm:
(IPC::ArgumentCoder<WebCore::AppKitControlSystemImage>::encode):
(IPC::ArgumentCoder<WebCore::AppKitControlSystemImage>::decode):

Canonical link: https://commits.webkit.org/257421@main
